File: backend/users/domain/repositories/user_repository.py
from abc import ABC, abstractmethod
from typing import List, Optional, Dict, Any
from ...models import User, Role # Importar Role

class IUserRepository(ABC):
    """
    Interfaz que define el contrato para un repositorio de usuarios.
    Esta interfaz reside en la capa de Dominio.
    """

    @abstractmethod
    async def get_by_id(self, user_id: int) -> Optional[User]:
        """
        Obtiene un usuario específico por su ID.
        """
        pass

    @abstractmethod
    async def get_by_username(self, username: str) -> Optional[User]:
        """
        Obtiene un usuario específico por su nombre de usuario.
        """
        pass
    
    @abstractmethod
    async def get_by_email(self, email: str) -> Optional[User]:
        """
        Obtiene un usuario específico por su email.
        """
        pass

    @abstractmethod
    async def get_role_by_name(self, role_name: str) -> Optional[Role]:
        """
        Obtiene un objeto Role por su nombre.
        """
        pass

    @abstractmethod
    async def create(self, user_data: Dict[str, Any]) -> User:
        """
        Crea un nuevo usuario.
        user_data debe incluir 'username', 'email', 'password'.
        Opcionalmente 'first_name', 'last_name', 'age'.
        """
        pass

    @abstractmethod
    async def update(self, user_id: int, user_data: Dict[str, Any]) -> Optional[User]:
        """
        Actualiza un usuario existente.
        """
        pass

    @abstractmethod
    async def get_all(self, filters: Optional[Dict[str, Any]] = None) -> List[User]:
        """
        Obtiene una lista de todos los usuarios, con filtros opcionales (ej. por rol).
        """
        pass

    @abstractmethod
    async def delete(self, user_id: int) -> bool:
        """
        Elimina un usuario existente. Devuelve True si se eliminó, False en caso contrario.
        """
        pass
    
    # La lógica de login/autenticación se maneja a través de IAuthRepository
    # y los endpoints de JWT/dj-rest-auth.
    # Si se necesita una lógica de usuario más compleja (ej. cambiar contraseña, verificar email),
    # se pueden añadir métodos aquí.

File: frontend/src/domain/repositories/user-repository.js
// import { User } from '../entities/user'; // Asumiendo que existe una entidad User en el frontend

/**
 * Interfaz que define el contrato para un repositorio de usuarios en el frontend.
 * Esta interfaz reside en la capa de Dominio y no debe depender de capas externas.
 */
export class IUserRepository {
  /**
   * Obtiene una lista de todos los usuarios, con filtros opcionales.
   * @param {object} [filters] - Filtros opcionales (ej. { role: 'admin' }).
   * @returns {Promise<object[]>} Una promesa que resuelve con un array de objetos de usuario.
   *                              (Debería ser Promise<User[]> si se define la entidad User en frontend)
   */
  async getAllUsers(filters) {
    throw new Error('Method not implemented.');
  }

  /**
   * Crea un nuevo usuario (admin de cancha).
   * @param {object} userData - Datos del usuario a crear.
   * @returns {Promise<object>} Una promesa que resuelve con el objeto del usuario creado.
   */
  async createAdminUser(userData) {
    throw new Error('Method not implemented.');
  }

  /**
   * Actualiza el estado (activo/inactivo) de un usuario.
   * @param {number} userId - El ID del usuario.
   * @param {boolean} isActive - El nuevo estado de activación.
   * @returns {Promise<object>} Una promesa que resuelve con el objeto del usuario actualizado.
   */
  async updateUserStatus(userId, isActive) {
    throw new Error('Method not implemented.');
  }

  /**
   * Elimina un usuario.
   * @param {number} userId - El ID del usuario.
   * @returns {Promise<void>} Una promesa que resuelve cuando el usuario ha sido eliminado.
   */
  async deleteUser(userId) {
    throw new Error('Method not implemented.');
  }
  
  // Otros métodos relacionados con usuarios si son necesarios
}
